---
title: "Experimental merge reads"
author: "Rob Harbert"
date: "6/21/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Merging paired end reads in R

# Dada2 to save it all

```{R}
#install dada2 with: library("devtools")
devtools::install_github("benjjneb/dada2", ref="v1.12")

library(dada2)
#derepF = derepFastq(system.file("extdata", "sam1F.fastq.gz", package="dada2"))
#derepR = derepFastq(system.file("extdata", "sam1R.fastq.gz", package="dada2"))
path2R1 = 'test.R1.fastq'
path2R2 =  'test.R2.fastq'
derepF = derepFastq(path2R1)
derepR = derepFastq(path2R2)
dadaF <- dada(derepF, err=tperr1, errorEstimationFunction=loessErrfun, selfConsist=TRUE)
dadaR <- dada(derepR, err=tperr1, errorEstimationFunction=loessErrfun, selfConsist=TRUE)
merger.basic <- mergePairs(dadaF, derepF, dadaR, derepR)  # probably use this
merger.withrejects <- mergePairs(dadaF, derepF, dadaR, derepR, returnRejects=TRUE, propagateCol=c("n0", "birth_ham"))
merger.concat <- mergePairs(dadaF, derepF, dadaR, derepR, justConcatenate=TRUE)

# create sequence table:
seqtab <- makeSequenceTable(merger.basic)
dim(seqtab)
# remove chimeras
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)

# assign taxonomy (Naive Bayes classifier -- Look for more on constructing reference fasta -- This does not work with a plain fasta file)
#taxa <- assignTaxonomy(seqtab.nochim, "/path/to/reference", multithread=TRUE)
#taxa <- addSpecies(taxa, "~/tax/silva_species_assignment_v132.fa.gz")

#or classify with rBLAST to NCBI nt or custom DB

```
